tests/integration/component_provider/python/component-provider-host/provider/component.py

```diff
@@ -18,14 +18,25 @@
 import pulumi
 
 
+class Nested(TypedDict):
+    nested_str: str
+
+
+class Complex(TypedDict):
+    complex_str: pulumi.Input[str]
+    nested: pulumi.Input[Nested]
```

Comment:
Suggested test cases:
- A type that recursively references itself
- A recursive loop of type references

These do happen in existing providers and our schema/codegen knows how to handle those.

Reply:
Fun! I updated the PR for that.

Self recursive works nicely.

For mutually recursive types, it is a bit tricky, because you have to use a `ForwardRef` in Python, which tracks the type by name. I think my solution works reasonably well for common cases, but you could potentially construct types that it fails with right now, mostly if you declare types outside of the global scope, which is possible, but unlikely (I hope?).